include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -709,13 +731,344 @@ namespace Differentiation
+       * the values set with register_dof_values().
+       *
+       * @return An object of auto-differentiable type numbers representing the
+       * local degree-of-freedom values.
+       *
+       * @note This function is not typically used within the context of automatic
+       * differentation computations, but can make performing substutitions in
+       * other post-processing computations more straight forward.
+       *
+       * @note For taped AD numbers, this operation is only valid outside recording mode.
+       */
+      std::vector<ad_type>
+      get_non_sensitive_dof_values() const;
+
+      //@}
+
+      /**
+       * @name Operations specific to taped mode: Reusing tapes
+       */
+      //@{
+
+      /**
+       * Set the values for the independent variables $\mathbf{X}$, i.e. the
+       * linearization point.
+       *
+       * @param[in] dof_values A vector field associated with local
+       * degree-of-freedom values on the current finite element. These define
+       * the values of all independent variables.
+       *
+       * @note The input value type must correspond to this class's number type.
+       * So if this class is templated on type double then the number type
+       * associated with value must also be of type double.
+       *
+       * @note If the keep flag has been set when start_recording_operations() is
+       * called then the tape is immediately usable after creation, and the
+       * values of the independent variables set by register_dof_values() are
+       * those at which the function is to be evaluated. In this case, a call to
+       * this function is not strictly necessary.
+       */
+      void
+      set_dof_values(const std::vector<scalar_type> &dof_values);
+
+      /**
+       * Set the values for the independent variables $\mathbf{X}$, i.e. the
+       * linearization point.
+       *
+       * @param[in] values A vector field from which the values of all
+       * independent variables is to be extracted.
+       * @param[in] local_dof_indices A vector of degree-of-freedom indices from
+       * which to extract the local degree-of-freedom values.
+       *
+       * @note If the keep flag has been set when start_recording_operations() is
+       * called then the tape is immediately usable after creation, and the
+       * values of the independent variables set by register_dof_values() are
+       * those at which the function is to be evaluated. In this case, a call to
+       * this function is not strictly necessary.
+       */
+      template <typename VectorType>
+      void
+      set_dof_values(
+        const VectorType &                                  values,
+        const std::vector<dealii::types::global_dof_index> &local_dof_indices);
+
+      //@}
+
+      /**
+       * @name Dependent variables
+       */
+      //@{
+
+      /**
+       * Computes the value of the residual vector field
+       * $\mathbf{r}(\mathbf{X})$.
```

Comment:
I guess this function does not actually compute anything, but it should do so in derived classes?

It would probably be useful to state how large the vector is supposed to be that's returned. Also, we generally return `Vector` objects through arguments, because they're reused from quadrature point to quadrature point.

Same for the next function.

Reply:
> we generally return Vector objects through arguments

Well that change going to break a *few* tests...

Reply:
Actually, I'm not happy with making this change with the current class interface. 

The problem is that the size of the returned objects (for `ADHelperVariationalFormulation` and `ADHelperResidualLinearisation`) differ. For the former, the residual vector has length `n_independent_variables` (because its the first derivative of a scalar function, the only dependent variable) and the latter has length `n_dependent_variables` . 

The only way I see this working (i.e. having the output passed as an argument) is if we have another set of virtual functions that return the expected sizes of these objects. If you like this idea then do you have a suggestion for a name (I think something like `m()` and `n()` is not descriptive, and maybe `n_rows_residual`, `n_rows_linearization` and `n_cols_linearization` is too verbose.

Reply:
If you feel strongly, then let's leave it as is.

The argument for returning via an argument is that it doesn't require memory allocation for every call, whereas when you return by value, you'll have to allocate an object. I can't say for sure whether that's a noticeable overhead, but we've generally followed this strategy.